### DISCUSSION TITLE
How can I get the default texture uri from a material?

### DISCUSSION THREAD

**Discussion by discodevl** [google/model-viewer#2957] [Q&A] [answered]
Hi, I need to get each default texture from the materials of the model viewer, I managed to get them once I set the texture with material.['texture'].texture.source.setURI() method, before setting the uri it gives me undefined. How can I get this uri from the default texture of the modelViewer?

**Comment by ghost** [comment, marked as answer]
Textures loaded from a .glb have no URI because they are loaded from a buffer inside the .glb so there is no way to address them directly. You can check the type of the texture (`modelviewer.model.materials[0].normalTexture.source.type` for instance) if it's 'embedded' there will be no URI, and if it's 'external' there should be a defined URI.

What are you aiming to do? I might be able to provide some other suggestions with a little more detail.

**Reply by discodevl** [reply]
I want to display a thumbnail for each texture of the model-viewer material. For instance, after uploaded a glb file, I have 5 `<img src={}>` tags (one to display the base color texture, one to the metallic roughness texture, emissive texture, normal and occlusion)  that I want to pass the blob of the default textures of the glb's first material inside these src. Im doing like this:
```
let material = modelviewer.model.materials[0];
let blob = material.emissiveTexture.texture.source.uri;

<img src={blob}  alt="Emissive Texture" />
```
Result:
![emissiveBroke](https://user-images.githubusercontent.com/69987663/140421290-514b785e-3707-4324-a9ef-2da355a55a02.png)
Im aiming something like this:
![emissiveEx](https://user-images.githubusercontent.com/69987663/140421572-803a517c-7142-49db-bf60-1740a366277f.png)

**Reply by ghost** [reply]
Ah we have support for that with `createThumbnail()` .

Example:

```
let material = modelviewer.model.materials[0];
let thumbUrl = await material.emissiveTexture.texture.source.createThumbnail(128, 128);
<img src={thumbUrl}  alt="Emissive Texture" />
```

**Reply by discodevl** [reply]
It works :D
Thank you!

**Reply by canberka** [reply]
Hi, i tried these codes but not worked. I am failing somewhere hard but not get it where is it ? Can you help me? 

```
<img id="thumbnail" src={thumbUrl} />
<script>
  const modelViewerTexture = document.querySelector("model-viewer#sphere");
  modelViewerTexture.addEventListener("load", (ev) => {
  let material = modelViewerTexture.model.materials[7];
   let thumbUrl =await material.baseColorTexture.texture.source.createThumbnail(128, 128);
})
</script>
```
this one" Uncaught SyntaxError: await is only valid in async functions and the top level bodies of modules"
```

<script>
 const modelViewerTexture = document.querySelector("model-viewer#sphere");
modelViewerTexture.addEventListener("load", (ev) => {
   (async function(){
let material = modelViewerTexture.model.materials[7];
    let thumbUrl =await material.baseColorTexture.texture.source.createThumbnail(128, 128);
})})

</script>
```
this one get nothing


```
<img id="thumbnail" src={thumbUrl} />
<script>
const modelViewerTexture = document.querySelector("model-viewer#sphere");
modelViewerTexture.addEventListener("load", (ev) => {
  let material = modelViewerTexture.model.materials[7];
 let thumbUrl =material.baseColorTexture.texture.source.createThumbnail(128, 128);
})

</script>
```
this one " Uncaught TypeError: Cannot read properties of undefined (reading 'texture')"

**Reply by elalish** [reply]
This is just JS debug; you may  need to read up on how `async` and `await` are used. The last error makes it  sound like your eighth material has no baseColor texture. You might try opening this GLB in our editor to see if that gives you any insight.

**Reply by canberka** [reply]
thanks for quick response i will read up asap ... if it is possible to help @discodevl  i really would be glad

last error in editor is @elalish 

![image](https://user-images.githubusercontent.com/36857390/144101723-5c9c03c9-5f62-4e3a-9bab-e89e4000a586.png)

**Reply by discodevl** [reply]
Im working with reactJS so maybe the exact same code wont work for you but I did this:

```

const material = document.querySelector('model-viewer').model.materials[0];

async function defaultThumb() {
    const promise = await material.pbrMetallicRoughness.baseColorTexture.texture.source.createThumbnail(256,256);
    return promise;
  }

```
I used the useEffect hook from react to assign to `source` the value of `defaultThumb()`.
The code inside useEffect will run once the page is rendered.
```
useEffect(async () => {
    const source = await defaultThumb();
  }, []);
```
Then I just have to put the `source` inside the `src` on the img tag.
```
<img src={source} />
```
Not the cleanest way to do it but it worked for me

**Comment by discodevl** [comment]
After doing that, I came up a doubt. How can I revert the texture to the default one properly? After changing the texture if I try to set the texture with `setURI()`  method passing `thumbUrl` the model comes pixelated, I tried to change the values of the `createThumbnail()`  but the result still bad. How can I 'revert' the texture?

**Reply by ghost** [reply]
A couple of things, you may need to make the 'thumbnail' larger, else it will be pixelated, you probably don't want to apply a small thumbnail back to your model. So if your source texture is 4096x4096 (or whatever it might be) then you want to go with that resolution.

To revert the texture you need to save a reference to the original texture.
```
const saveTexture = material.emissiveTexture.texture;
// revert later on
material.emissiveTexture.setTexture(saveTexture);
```

**Reply by buddy90210** [reply]
`> To revert the texture you need to save a reference to the original texture.`
I am saving the texture in const saveTexture and then after changing the texture of the material I want to return the texture to default, but I have an error Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'Symbol(threeTexture)').
Default texture is inside the .glb, when i change this texture i set external texture image.

**Reply by elalish** [reply]
@buddy90210 Can you share a repro URL? It's possible this is a bug.

**Reply by buddy90210** [reply]
Oh, sorry! I forgot add 'await' to my function.

**Reply by adesidaseyi** [reply]
For anyone working with Diffuse textures and trying to revert the base color texture of their model, following a similar approach to what was discussed earlier (saving a reference to the original texture object), you could do something like this:


```
// 'baseColorTexture' is the channel for Diffuse textures
const saveTexture = material.pbrMetallicRoughness['baseColorTexture'].texture; 

// revert later on
material.pbrMetallicRoughness['baseColorTexture'].setTexture(saveTexture);
```

**Comment by ColinCollins** [comment]
`createThumbnail` the result with different color space make texture become dark!
First serveral days, I still could got the source.uri, but when few days age, the datas beceme undefined. I didn't know what's wrong?
![eMVdl0iOHL](https://github.com/google/model-viewer/assets/35832931/c2f003d6-449e-4cc5-bee5-82cdb6e6b44f)

model-viewer: 2.1.1

**Reply by elalish** [reply]
I can't quite follow what you mean - can you show a comparison to what you would expect to see? Do you have a repro link? When you say something changed after several days, do you mean you changed the MV version?

**Comment by WadeMorrison** [comment]
Is it possible to revert a material using the originalGltfJson? I'm trying to make it to where if I swap to a specific material type, it will automatically swap another part back to it's original material because the newly "checked" material cannot be created/sold with that combination. 

Here is what I'm trying to achieve, I've got it to swap the material on DrawerBody if a lacquer color is chosen, but when I choose a Grasscloth material, it still swaps it to that material instead of back to the default "wood" material.

```
const swapDrawers = (event) => {
        const modelViewer = document.querySelector('model-viewer#model-viewer')
        const material = modelViewer.model.getMaterialByName('Drawer');
        const drawerBody = modelViewer.model.getMaterialByName('DrawerBody');
        const ogDiffuse= modelViewer.originalGltfJson.textures[4].source;
        const checkedMat = document.querySelector('[data-type]:checked').getAttribute('data-type')
  
          const diffuseSwap = async (channel) => {
            const diffuse = await modelViewer.createTexture(event.target.getAttribute('value'));
              if (channel.includes('base')) 
              {
                material.pbrMetallicRoughness[channel].setTexture(diffuse);
                if (checkedMat === 'Lacquer') {
                  drawerBody.pbrMetallicRoughness[channel].setTexture(diffuse);
                } else {
                  drawerBody.pbrMetallicRoughness[channel].setTexture(ogDiffuse);
                }
              }
            }
                
          const normalSwap = async (channel) => {
            const normal = await modelViewer.createTexture(event.target.getAttribute('data-value'));
              if (channel.includes('normal')) 
              {
                material[channel].setTexture(normal);
              }
            }
                
          const ORMSwap = async (channel) => {
            const ORM = await modelViewer.createTexture(event.target.getAttribute('data-value2'));
              if (channel.includes('metallic')) 
              {
                material.pbrMetallicRoughness[channel].setTexture(ORM);
              }
            }
  
        diffuseSwap('baseColorTexture');
        normalSwap('normalTexture');
        ORMSwap('metallicRoughnessTexture');

        if (checkedMat !== 'Lacquer') {
          drawerBody.pbrMetallicRoughness['baseColorTexture'].setTexture(ogDiffuse);
        }

        console.log("swapping material")
      };
     ```

**Reply by elalish** [reply]
Yes, we do basically exactly this in our editor. Feel free to look at the `materials_panel` code in our `space-opera` package.

**Reply by elalish** [reply]
It's all under the same open license, so you can just fork space-opera if you wanted to.

**Reply by WadeMorrison** [reply]
Okay I will take a look, I appreciate it!
